test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala

```diff
@@ -0,0 +1,173 @@
+package scala
+package tools.nsc
+package reporters
+
+import java.io.{ByteArrayOutputStream, StringReader, BufferedReader, PrintStream, PrintWriter}
+import org.junit.Assert._
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+import scala.reflect.internal.util._
+
+
+@RunWith(classOf[JUnit4])
+class ConsoleReporterTest {
+  val source = "Test_ConsoleReporter"
+  val batchFile = new BatchSourceFile(source, "For testing".toList)
+  val posWithSource = new OffsetPosition(batchFile, 4)
+  val content = posWithSource.lineContent
+  val writerOut = new ByteArrayOutputStream()
+  val echoWriterOut = new ByteArrayOutputStream()
+
+
+  def createConsoleReporter(inputForReader: String, errOut: ByteArrayOutputStream, echoOut: ByteArrayOutputStream = null): ConsoleReporter = {
+    val reader = new BufferedReader(new StringReader(inputForReader))
+      
+    /** Create reporter with the same writer and echoWriter if echoOut is null */
+    echoOut match {
+      case null => new ConsoleReporter(new Settings(), reader, new PrintWriter(errOut))
+      case _ => new ConsoleReporter(new Settings(), reader, new PrintWriter(errOut), new PrintWriter(echoWriterOut))
+    }
+  }
+
+
+  def testHelper(pos: Position = NoPosition, msg: String, severity: String = "")(test: Position => Unit) = {
+    test(pos)
+    if (msg.isEmpty && severity.isEmpty) assertTrue(writerOut.toString.isEmpty)
+    else {
+      if (!pos.isDefined) assertEquals(severity + msg, writerOut.toString.lines.next)
+      else {
+        val it = writerOut.toString.lines
+        assertEquals(source + ":1: " + severity + msg, it.next)
+        assertEquals(content, it.next)
+        assertEquals("    ^", it.next)
+      }
+    }
+    writerOut.reset
```

Comment:
Technically, we should have put the reset in a finally block. If it fails, it throws, so the next test to run could see the dirty buffer. Probably the test class is reinstantiated; I think junit does that.